File: src/Tests/Behavioral/GenericInterfaceConstraint/GenericInterfaceConstraint.go
// Guards the generic-over-interface constraint lowering: a type parameter constrained
// by a REGULAR method-set interface (`totalArea[S Shape]`) emits `where S : Shape`
// against the arity-0 interface (NOT the phantom CRTP `Shape<S>, new()` — CS0308), and
// a call site instantiating the type parameter with a POINTER type projects the slice
// element-wise through the generated pointer adapter (`widen<ж<Circle>, Shape>(...)`),
// since the `ж<Circle>` box does not implement the interface — its adapter does.
// Mirrors go/ast's `walkList[N Node](v Visitor, list []N)`.
package main

import "fmt"

type Shape interface {
	Area() float64
	Name() string
}

// Round embeds Shape — an interface type argument through inheritance mirrors
// ast.Stmt/Expr/Spec/Decl instantiating walkList's N.
type Round interface {
	Shape
	Diameter() float64
}

type Circle struct {
	R float64
}

func (c *Circle) Area() float64 {
	return 3.0 * c.R * c.R
}

func (c *Circle) Name() string {
	return "circle"
}

func (c *Circle) Diameter() float64 {
	return 2.0 * c.R
}

type Square struct {
	S float64
}

func (s *Square) Area() float64 {
	return s.S * s.S
}

func (s *Square) Name() string {
	return "square"
}

// totalArea calls a constraint method ON the type parameter (richer than walkList).
func totalArea[S Shape](shapes []S) float64 {
	var sum float64
	for _, s := range shapes {
		sum += s.Area()
	}
	return sum
}

// walkAll mirrors go/ast's walkList exactly: the type parameter only WIDENS to the
// interface inside the body.
func walkAll[S Shape](shapes []S) {
	for _, s := range shapes {
		show(s)
	}
}

func show(s Shape) {
	fmt.Printf("%s: %.2f\n", s.Name(), s.Area())
}

func main() {
	circles := []*Circle{&Circle{R: 1}, &Circle{R: 2}} // pointer instantiation — adapter-projected
	squares := []*Square{&Square{S: 3}}                // second pointer type — distinct adapter
	shapes := []Shape{&Circle{R: 1}, &Square{S: 2}}    // interface instantiation — direct
	rounds := []Round{&Circle{R: 4}}                   // embedded-interface instantiation

	fmt.Printf("circles: %.2f\n", totalArea(circles))
	fmt.Printf("squares: %.2f\n", totalArea(squares))
	fmt.Printf("shapes: %.2f\n", totalArea(shapes))
	fmt.Printf("rounds: %.2f\n", totalArea(rounds))

	walkAll(circles)
	walkAll(shapes)
	walkAll(rounds)
}

File: src/Tests/Behavioral/GenericInterfaceConstraint/GenericInterfaceConstraint.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial interface Shape {
    float64 Area();
    @string Name();
}

[GoType] partial interface Round :
    Shape
{
    float64 Diameter();
}

[GoType] partial struct Circle {
    public float64 R;
}

[GoRecv] public static float64 Area(this ref Circle c) {
    return 3.0D * c.R * c.R;
}

[GoRecv] public static @string Name(this ref Circle c) {
    return "circle"u8;
}

[GoRecv] public static float64 Diameter(this ref Circle c) {
    return 2.0D * c.R;
}

[GoType] partial struct Square {
    public float64 S;
}

[GoRecv] public static float64 Area(this ref Square s) {
    return s.S * s.S;
}

[GoRecv] public static @string Name(this ref Square s) {
    return "square"u8;
}

internal static float64 totalArea<S>(slice<S> shapes)
    where S : Shape
{
    float64 sum = default!;
    foreach (var (_, s) in shapes) {
        sum += s.Area();
    }
    return sum;
}

internal static void walkAll<S>(slice<S> shapes)
    where S : Shape
{
    foreach (var (_, s) in shapes) {
        show(s);
    }
}

internal static void show(Shape s) {
    fmt.Printf("%s: %.2f\n"u8, s.Name(), s.Area());
}

internal static void Main() {
    var circles = new ж<Circle>[]{Ꮡ(new Circle(R: 1D)), Ꮡ(new Circle(R: 2D))}.slice();
    var squares = new ж<Square>[]{Ꮡ(new Square(S: 3D))}.slice();
    var shapes = new Shape[]{new CircleжShape(Ꮡ(new Circle(R: 1D))), new SquareжShape(Ꮡ(new Square(S: 2D)))}.slice();
    var rounds = new Round[]{new CircleжRound(Ꮡ(new Circle(R: 4D)))}.slice();
    fmt.Printf("circles: %.2f\n"u8, totalArea(widen<ж<Circle>, Shape>(circles, elemᴛ0 => new CircleжShape(elemᴛ0))));
    fmt.Printf("squares: %.2f\n"u8, totalArea(widen<ж<Square>, Shape>(squares, elemᴛ0 => new SquareжShape(elemᴛ0))));
    fmt.Printf("shapes: %.2f\n"u8, totalArea(shapes));
    fmt.Printf("rounds: %.2f\n"u8, totalArea(rounds));
    walkAll(widen<ж<Circle>, Shape>(circles, elemᴛ0 => new CircleжShape(elemᴛ0)));
    walkAll(shapes);
    walkAll(rounds);
}

} // end main_package
